Add Assert.Throws/ThrowsAny analyzer

ErikSchierboom opened the pull request:
Refs https://github.com/xunit/xunit/issues/1248

ErikSchierboom committed:
Add Assert.Throws/ThrowsAny analyzer

dnfclas commented:
@ErikSchierboom,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

marcind commented:
The generic versions of Assert.Throws have overrides that match a Task-returning method that are marked as obsolete with error, so the following two usages are already marked as compilation errors:

```C#
Assert.Throws<Exception>(() => Foo());
Assert.Throws<ArgumentException>("foo", () => Foo());
```

In those cases the analyzer should not raise an additional warning but just provide a refactoring to fix the problem.

The 3rd handled case `Assert.Throws(typeof(Exception), () => Foo());` does not have an obsoleted overload for some reason, but causes a runtime error thus in this case the analyzer should produce an error.

ErikSchierboom commented:
Ah good point. How can I best do that?

marcind commented on src/xunit.analyzers/AssertThrowsShouldNotBeUsedForAsyncThrowsCheckFixer.cs:
```diff
@@ -0,0 +1,156 @@
+            if (taskType.IsAssignableFrom(methodSymbol))
+                return method.ReturnType;
+
+            return
```
This implementation builds a return type expression that's fully namespace qualified, even if the document already has the `System.Threading.Tasks` namespace included. I've used `editor.Generator.TypeExpression` to get something takes the imported namespaces into account.

ErikSchierboom replied:
Thanks for the tip! I didn't know how to do that, so that's why I used the fully qualified namespace.

marcind commented:
There's `Microsoft.CodeAnalysis.CodeRefactorings.CodeRefactoringProvider`, but I have not explored it that much. It could also probably be done using one descriptor with a severity of hidden (see X2001_AssertEqualsShouldNotBeUsed) and one with an Error.

ErikSchierboom commented:
@marcind I've updated the PR with the following changes:

- The two obsolete methods are not flagged as warnings anymore but use the hidden severity.
- I've added unit tests for the paramName overload (forgot about them).
- I've used `editor.Generator.TypeExpression` to generate the minimal `Task` return type syntax.
- I've rebased on the latest `master`.

marcind commented on src/xunit.analyzers/Descriptors.cs:
```diff
@@ -220,6 +220,16 @@ internal static class Descriptors
+        internal static DiagnosticDescriptor X2014_AssertThrowsShouldNotBeUsedForAsyncThrowsCheck { get; } = new DiagnosticDescriptor("xUnit2014",
+            "Do not use throws check to check for asynchronously thrown exception",
+            "Do not use {0} to check for asynchronously thrown exceptions.",
+            Categories.Assertions, DiagnosticSeverity.Warning, isEnabledByDefault: true);
```
This should be Error, since not addressing the problem will cause the test to fail at runtime.

ErikSchierboom replied:
Fixed!

ErikSchierboom commented:
Thanks for the thorough reviews!
